source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+#include "common/http/filter/gzip_filter.h"
+
+#include <regex>
+
+#include "common/common/macros.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Http {
+
+namespace {
+// Default zlib memory level.
+const uint64_t DefaultMemoryLevel{5};
+
+// Default and maximum compression window size.
+const uint64_t DefaultWindowBits{12};
+
+// Minimum length of an upstream response that allows compression.
+const uint64_t MinimumContentLength{30};
+
+// When summed to window bits, this sets a gzip header and trailer around the compressed data.
+const uint64_t GzipHeaderValue{16};
+
+// Used for verifying accept-encoding values.
+const char ZeroQvalueString[] = "q=0";
+
+// Default content types will be used if any is provided by the user.
+const std::unordered_set<std::string> defaultContentEncoding() {
+  CONSTRUCT_ON_FIRST_USE(std::unordered_set<std::string>,
+                         {"text/html", "text/plain", "text/css", "application/javascript",
+                          "application/json", "image/svg+xml", "text/xml",
+                          "application/xhtml+xml"});
+}
+
+} // namespace
+
+GzipFilterConfig::GzipFilterConfig(const envoy::api::v2::filter::http::Gzip& gzip)
+    : compression_level_(compressionLevelEnum(gzip.compression_level())),
+      compression_strategy_(compressionStrategyEnum(gzip.compression_strategy())),
+      content_length_(contentLengthUint(gzip.content_length().value())),
+      memory_level_(memoryLevelUint(gzip.memory_level().value())),
+      window_bits_(windowBitsUint(gzip.window_bits().value())),
+      content_type_values_(contentTypeSet(gzip.content_type())),
+      disable_on_etag_header_(gzip.disable_on_etag_header()),
+      disable_on_last_modified_header_(gzip.disable_on_last_modified_header()) {}
+
+ZlibCompressionLevelEnum
+GzipFilterConfig::compressionLevelEnum(const GzipV2CompressionLevelEnum& compression_level) {
+  switch (compression_level) {
+  case GzipV2CompressionLevelEnum::Gzip_CompressionLevel_Enum_BEST:
+    return ZlibCompressionLevelEnum::Best;
+  case GzipV2CompressionLevelEnum::Gzip_CompressionLevel_Enum_SPEED:
+    return ZlibCompressionLevelEnum::Speed;
+  default:
+    return ZlibCompressionLevelEnum::Standard;
+  }
+}
+
+ZlibCompressionStrategyEnum GzipFilterConfig::compressionStrategyEnum(
+    const GzipV2CompressionStrategyEnum& compression_strategy) {
+  switch (compression_strategy) {
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_RLE:
+    return ZlibCompressionStrategyEnum::Rle;
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_FILTERED:
+    return ZlibCompressionStrategyEnum::Filtered;
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_HUFFMAN:
+    return ZlibCompressionStrategyEnum::Huffman;
+  default:
+    return ZlibCompressionStrategyEnum::Standard;
```

Comment:
I'm not sure that we can leave default out. I think the compiler will complain. It's still valid to add a return after the switch in case this function receives an unexpected value tho.

Reply:
My bad...I wasn't conscious here that this came from a proto enum, where that's pretty much impossible to do.

Reply:
Ok, I will remove the last return in the next round :)